### PR TITLE
osd: subscribe osdmaps if any pending pgs

### PR DESCRIPTION
**Pull request by tchaikov**
subscribe from monitor continously for new osdmaps so osd is able to get
the osdmap instructing it to delete PGs even if nobody is poking it.

Fixes: http://tracker.ceph.com/issues/22113
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -4217,14 +4218,40 @@ void OSD::resume_creating_pg()
...
+	      << start << dendl;
+      do_renew_subs = true;
+    }
+  } else if (pgtemp || do_sub_pg_creates) {
+    // no need to subscribe the osdmap continuously anymore
+    // once the pgtemp and/or mon_subscribe(pg_creates) is sent
+    if (monc->sub_want_increment("osdmap", start, CEPH_SUBSCRIBE_ONETIME)) {
+      dout(4) << __func__ << ": re-subscribe osdmap(onetime) since"
+	      << start << dendl;
+      do_renew_subs = true;
```
Comment (liewegas) [inline src/osd/OSD.cc]: Once we have pending creates and a full subscription, how does it stop the subscription after the pg gets created?

Comment (tchaikov) [reply src/osd/OSD.cc]: @liewegas that's what https://github.com/ceph/ceph/pull/18916/files#diff-fa6c2eba8356ae1442d1bf749beacfdfR4244 is for.

Comment (liewegas) [reply src/osd/OSD.cc]: what if !(pgtemp || do_sub_pg_creates) thought?

Comment (tchaikov) [reply src/osd/OSD.cc]: @liewegas let me rephrase the code:
```c++
if (have_pending_creates) {
  subscribe(osdmap, continuously);
}
if (!have_pending_create) {
  if (pgtemp || do_sub_pg_creates) {
     /// the reason why we don't have pending creates is that
     /// we are about to create the pending pgs, and that means
     /// we've being subscribing the osdmap continuously, so 
     subscribe(osdmap, onetime); 
  } else {
     /// awesome, no pending creates now, and no pending creates **before** calling resume_creating_pg()
  }
}
```

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -4217,14 +4218,40 @@ void OSD::resume_creating_pg()
       pg = pending_creates_from_osd.erase(pg);
       spare_pgs--;
     }
+    have_pending_creates = (pending_creates_from_mon > 0 ||
+			    !pending_creates_from_osd.empty());
   }
+
+  bool do_renew_subs = false;
   if (do_sub_pg_creates) {
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/osd/OSD.cc]: it's set if `pending_creates_from_mon > 0`. it's not part of this patch.

[the remark this answers is not in the document]
Comment (songbaisen) [reply src/osd/OSD.cc]: oh, got it!

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-11-18_15:01:48-rados-wip-kefu-testing-2017-11-18-2124-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-11-20_05:45:10-rados-wip-kefu-testing-2017-11-18-2124-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-11-20_06:28:09-rados-wip-kefu-testing-2017-11-18-2124-distro-basic-mira/
